### PR TITLE
初步完成ADT

### PR DESCRIPTION
**Pull request by HancCui**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Sorry, No Test

**Test Configuration**:
* Operating System:
* Hardware:
* SDK Version:

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: OpenSchedule/Course.cs [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    {
+        private readonly ISet<CourseInformation> CourseSet = new SortedSet<CourseInformation>();
+
+        private readonly ISet<ExamInformation> ExamSet = new SortedSet<ExamInformation>();
+
+        /// <summary>
+        ///     Constructor
+        /// </summary>
+        /// <param name="name">the name of the course</param>
+        public Course(string name, string IdNum)
```
Comment (mahoshojoHCG) [inline OpenSchedule/Course.cs]: IdNum是什么？
- 不符合形式参数命名规范
- 没有XML注释

File: OpenSchedule/Course.cs [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    public class Course
+    {
+        private readonly ISet<CourseInformation> CourseSet = new SortedSet<CourseInformation>();
+
+        private readonly ISet<ExamInformation> ExamSet = new SortedSet<ExamInformation>();
+
+        /// <summary>
+        ///     Constructor
+        /// </summary>
+        /// <param name="name">the name of the course</param>
```
Comment (mahoshojoHCG) [inline OpenSchedule/Course.cs]: the name of the course表意不清，请考虑 Name of the course.

File: OpenSchedule/Course.cs [hunk 3]
```diff
@@ -0,0 +1,101 @@
+﻿using System.Collections.Generic;
```
Comment (mahoshojoHCG) [inline OpenSchedule/Course.cs]: 头部没有添加Apache协议文本，所有源代码都有这个问题。

File: OpenSchedule/Course.cs [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+        /// </summary>
+        private string CourseName { get; }
+
+        private string CourseId { get; }
+
+        /// <summary>
+        ///     get the name of the course
+        /// </summary>
+        /// <returns>the name of the course</returns>
+        public string GetCourseName()
```
Comment (mahoshojoHCG) [inline OpenSchedule/Course.cs]: 这是什么？

File: OpenSchedule/Course.cs [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+        /// <summary>
+        ///     A course, containing a set of eventInformation(including courseInformation and examInformation)
+        ///     一门课程，实际为若干课程信息和考试信息的集合
+        /// </summary>
+        private string CourseName { get; }
+
+        private string CourseId { get; }
+
+        /// <summary>
+        ///     get the name of the course
```
Comment (mahoshojoHCG) [inline OpenSchedule/Course.cs]: 英文请使用大写开头

File: OpenSchedule/Course.cs [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+        /// <param name="name">the name of the course</param>
+        public Course(string name, string IdNum)
+        {
+            CourseName = name;
+            CourseId = IdNum;
+        }
+
+        /// <summary>
+        ///     A course, containing a set of eventInformation(including courseInformation and examInformation)
+        ///     一门课程，实际为若干课程信息和考试信息的集合
```
Comment (mahoshojoHCG) [inline OpenSchedule/Course.cs]: 请不要中英文混搭，不同语言请写入单独的XML

File: OpenSchedule/Course.cs [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+        /// </summary>
+        private string CourseName { get; }
+
+        private string CourseId { get; }
```
Comment (mahoshojoHCG) [inline OpenSchedule/Course.cs]: 没有XML注释

File: OpenSchedule/Course.cs (hunk 6 above)
Comment (mahoshojoHCG) [reply OpenSchedule/Course.cs]: 而且注释门不对题

File: OpenSchedule/Course.cs (hunk 4 above)
Comment (mahoshojoHCG) [reply OpenSchedule/Course.cs]: 请使用属性

File: OpenSchedule/Course.cs [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+        ///     get the name of the course
+        /// </summary>
+        /// <returns>the name of the course</returns>
+        public string GetCourseName()
+        {
+            return CourseName;
+        }
+
+        /// <summary>
+        ///     get the Id of the Course
```
Comment (mahoshojoHCG) [inline OpenSchedule/Course.cs]: 请修改注释内容

File: OpenSchedule/Course.cs [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+        /// </summary>
+        /// <returns>the Id of the Course</returns>
+        public string GetCourseId()
```
Comment (mahoshojoHCG) [inline OpenSchedule/Course.cs]: 请使用属性

File: OpenSchedule/Course.cs [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+        {
+            return CourseId;
+        }
+
+        /// <summary>
+        ///     add a new course event to the course
```
Comment (mahoshojoHCG) [inline OpenSchedule/Course.cs]: 请修改注释以匹配英文语法，下同，以后不再提醒

File: OpenSchedule/CourseInformation.cs [hunk 11]
```diff
@@ -0,0 +1,28 @@
+﻿using System;
```
Comment (mahoshojoHCG) [inline OpenSchedule/CourseInformation.cs]: 头部没有添加Apache协议文本，所有源代码都有这个问题。

File: OpenSchedule/CourseInformation.cs [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+        /// <param name="teacherName">the name of the teacher of the course</param>
+        public CourseInformation(string room, DateTime start, TimeSpan duration, string teacherName) : base(room, start,
+            duration)
+        {
+            Teacher = teacherName;
+        }
+
+        /// <summary>
+        ///     The info of a course,including the teacher's name,and the info of the event(classroom,start time and the duration)
+        ///     继承自类EventInformation，表示一个课程。内容包括：教师名称 Teacher，以及父类的变量Classroom, StartTime, Duration
```
Comment (mahoshojoHCG) [inline OpenSchedule/CourseInformation.cs]: 语言问题

File: OpenSchedule/EventInformation.cs [hunk 13]
```diff
@@ -0,0 +1,79 @@
+﻿using System;
```
Comment (mahoshojoHCG) [inline OpenSchedule/EventInformation.cs]: 头部没有添加Apache协议文本，所有源代码都有这个问题。

File: OpenSchedule/EventInformation.cs [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+
+namespace OpenSchedule
+{
+    public class EventInformation : IComparable, ICloneable
+    {
+        /// <summary>
+        ///     the constructor of class EventInformation
```
Comment (mahoshojoHCG) [inline OpenSchedule/EventInformation.cs]: the constructor

File: OpenSchedule/EventInformation.cs [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+        public EventInformation(string room, DateTime start, TimeSpan duration)
+        {
+            Classroom = room;
+            StartTime = start;
+            EventDuration = duration;
+            EventId = Guid.NewGuid();
+        }
+
+        /// <summary>
+        ///     a new constructor of class EventInformation, only used in method clone()
```
Comment (mahoshojoHCG) [inline OpenSchedule/EventInformation.cs]: 为什么这里又用a了，请仔细检查英文语法

File: OpenSchedule/Course.cs (hunk 1 above)
Comment (HancCui) [reply OpenSchedule/Course.cs]: 指这一个变量不满足下面两点，还是全部变量？

File: OpenSchedule/EventInformation.cs [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+        /// </summary>
+        /// <param name="room"></param>
+        /// <param name="start"></param>
+        /// <param name="duration"></param>
+        /// <param name="Id"></param>
+        public EventInformation(string room, DateTime start, TimeSpan duration, Guid Id)
```
Comment (mahoshojoHCG) [inline OpenSchedule/EventInformation.cs]: 如果你只打算再Clone()方法中使用，可以考虑使用私有的构造函数，而且只在一处使用的话可以直接写在Clone()方法里

File: OpenSchedule/EventInformation.cs [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+            Classroom = room;
+            StartTime = start;
+            EventDuration = duration;
+            EventId = Id;
+        }
+
+        /// <summary>
+        ///     The info of an event, including classroom(the location of a class),
+        ///     the start time of the course and the duration of the course.
+        ///     表示一个事件的信息（可以是课程，也可以是考试），内容包含:地点 classroom，开始时间 StartTime，时长 EventDuration
```
Comment (mahoshojoHCG) [inline OpenSchedule/EventInformation.cs]: 语言问题

File: OpenSchedule/EventInformation.cs [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+        public DateTime StartTime { get; }
+        public TimeSpan EventDuration { get; }
+
+        /// <summary>
+        ///     get a new Event same with this event
+        /// </summary>
+        /// <returns>an object same with the current event</returns>
+        public object Clone()
+        {
+            return new EventInformation(Classroom, StartTime, EventDuration, EventId);
```
Comment (mahoshojoHCG) [inline OpenSchedule/EventInformation.cs]: ICloneable返回的是object，请添加返回实例相同类型的方法，并且将改方法签名修改为ICloneable.Clone()

File: OpenSchedule/EventInformation.cs [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+        }
+
+        /// <summary>
+        ///     compare the input obj and this event by time
+        /// </summary>
+        /// <param name="obj">the object compared with the current event</param>
+        /// <returns>
+        ///     1 if the obj is null
```
Comment (mahoshojoHCG) [inline OpenSchedule/EventInformation.cs]: null请抛出异常

File: OpenSchedule/EventInformation.cs [hunk 20]
```diff
@@ -0,0 +1,79 @@
...
+        ///     positive if the obj is earlier than the current event
+        ///     negative if the obj is later than the current event
+        ///     0 if the times are the same
+        /// </returns>
+        public int CompareTo(object obj)
```
Comment (mahoshojoHCG) [inline OpenSchedule/EventInformation.cs]: 请同时实现ICompare\<T\>

File: OpenSchedule/ExamInformation.cs [hunk 21]
```diff
@@ -0,0 +1,17 @@
+﻿using System;
```
Comment (mahoshojoHCG) [inline OpenSchedule/ExamInformation.cs]: 头部没有添加Apache协议文本，所有源代码都有这个问题。

File: OpenSchedule/Course.cs [hunk 22]
```diff
@@ -0,0 +1,101 @@
...
+
+namespace OpenSchedule
+{
+    /// <summary>
+    /// </summary>
+    public class Course
+    {
+        private readonly ISet<CourseInformation> CourseSet = new SortedSet<CourseInformation>();
+
+        private readonly ISet<ExamInformation> ExamSet = new SortedSet<ExamInformation>();
```
Comment (mahoshojoHCG) [inline OpenSchedule/Course.cs]: 私有字段不满足命名规则，正确的应该是_courseSet, _examSet

File: OpenSchedule/Course.cs (hunk 7 above)
Comment (mahoshojoHCG) [reply OpenSchedule/Course.cs]: Id应使用GUID类型

File: OpenSchedule/Schedule.cs [hunk 23]
```diff
@@ -0,0 +1,80 @@
...
+        /// </summary>
+        /// <param name="name">the name of the student</param>
+        public Schedule(string name)
+        {
+            StudentName = name;
+        }
+
+        /// <summary>
+        ///     the schedule of a student, containing all the course of the student.
+        ///     某个人的课程表，是他所需要上的所有课程的集合。
```
Comment (mahoshojoHCG) [inline OpenSchedule/Schedule.cs]: 语言问题

File: OpenSchedule/Schedule.cs [hunk 24]
```diff
@@ -0,0 +1,80 @@
...
+        ///     constructor
+        /// </summary>
+        /// <param name="name">the name of the student</param>
+        public Schedule(string name)
+        {
+            StudentName = name;
+        }
+
+        /// <summary>
+        ///     the schedule of a student, containing all the course of the student.
```
Comment (mahoshojoHCG) [inline OpenSchedule/Schedule.cs]: 用户不一定是student

File: OpenSchedule/Schedule.cs [hunk 25]
```diff
@@ -0,0 +1,80 @@
...
+        ///     某个人的课程表，是他所需要上的所有课程的集合。
+        /// </summary>
+        private string StudentName { get; }
+
+        /// <summary>
+        ///     get the name of the student
+        /// </summary>
+        /// <returns>
+        ///     <the name of the student/ returns>
+        public string GetStudentName()
```
Comment (mahoshojoHCG) [inline OpenSchedule/Schedule.cs]: 请使用属性

**Review by mahoshojoHCG — CHANGES_REQUESTED** [review]
问题很大，请逐条修改

File: OpenSchedule/Course.cs (hunk 1 above)
Comment (mahoshojoHCG) [reply OpenSchedule/Course.cs]: 只看到了这一点

File: OpenSchedule/EventInformation.cs [hunk 26]
```diff
@@ -0,0 +1,126 @@
...
+using System;
+
+namespace OpenSchedule
+{
+    /// <summary>
+    ///     Represants a set of information of an event,
+    ///     including classroom(the location of a class),
+    ///     the start time of the course and the duration of the course.
+    /// </summary>
+    public class EventInformation : IComparable, ICloneable
```
Comment (mahoshojoHCG) [inline OpenSchedule/EventInformation.cs]: 考虑更改为abstract class

**Review by mahoshojoHCG — CHANGES_REQUESTED** [review]

File: OpenSchedule/EventInformation.cs [hunk 27]
```diff
@@ -0,0 +1,126 @@
...
+        /// </param>
+        /// <returns>
+        ///     Throw a NullReferenceException if obj is null
+        ///     Positive if the obj is earlier than the current event
+        ///     Negative if the obj is later than the current event
+        ///     0 if the two times are the same
+        /// </returns>
+        public int CompareTo(object obj)
+        {
+            if (obj == null) throw new NullReferenceException();
```
Comment (mahoshojoHCG) [inline OpenSchedule/EventInformation.cs]: null判断不需要，如果是null，他也不是EventInformation。

File: OpenSchedule/EventInformation.cs [hunk 28]
```diff
@@ -0,0 +1,126 @@
...
+
+            if (obj is EventInformation otherEventInformation)
+                return StartTime.CompareTo(otherEventInformation.StartTime);
+
+            throw new ArgumentException("Object is not a EventInformation");
```
Comment (mahoshojoHCG) [inline OpenSchedule/EventInformation.cs]: 考虑使用name的重载

**Review by mahoshojoHCG — CHANGES_REQUESTED** [review]

**Comment by mahoshojoHCG** [comment]
不要忘记添加单元测试

**Comment by FerdinandSu** [comment]
已经添加单元测试，并修正了一些不规范的内容

**Comment by FerdinandSu** [comment]
已经清理代码

**Review by mahoshojoHCG — APPROVED** [review]

**Pull request merged**